components/dada-validate/src/validate/validator.rs

```diff
@@ -150,7 +150,30 @@ impl<'me> Validator<'me> {
     }
 
     #[tracing::instrument(level = "debug", skip_all)]
-    pub(crate) fn validate_expr(&mut self, expr: syntax::Expr) -> validated::Expr {
+    pub(crate) fn validate_root_expr(&mut self, expr: syntax::Expr) -> validated::Expr {
+        let validated_expr = self.validate_expr(expr);
+        if let Some(return_type) = self.code.return_type {
+            if let validated::ExprData::Seq(exprs) = validated_expr.data(self.tables) {
+                if exprs.is_empty() {
+                    dada_ir::error!(return_type, "must return something",)
+                        .primary_label("because return type is not unit")
+                        .emit(self.db);
```

Comment:
Let's see if we can tweak this wording...

```suggestion
                    dada_ir::error!(return_type, "function body cannot be empty",)
                        .primary_label("because function is supposed to return something")
                        .emit(self.db);
```

Maybe this?